internal/controllers/configuration/object_references.go

```diff
@@ -0,0 +1,267 @@
+package configuration
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	extv1beta1 "k8s.io/api/extensions/v1beta1"
+	netv1 "k8s.io/api/networking/v1"
+	netv1beta1 "k8s.io/api/networking/v1beta1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/annotations"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane"
+	kongv1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1"
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1beta1"
+)
+
+func updateReferredObjects(
+	ctx context.Context, client client.Client, dataplaneClient *dataplane.KongClient, obj runtime.Object) error {
+	switch obj := obj.(type) {
+	case *corev1.Service:
+		return updateCoreV1ServiceReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *netv1.Ingress:
+		return updateNetV1IngressReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *netv1beta1.Ingress:
+		return updateNetV1beta1IngressReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *extv1beta1.Ingress:
+		return updateExtensionV1beta1IngressReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *kongv1.KongPlugin:
+		return updateKongPluginReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *kongv1.KongClusterPlugin:
+		return updateKongClusterPluginReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *kongv1.KongConsumer:
+		return updateKongConsumerReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *kongv1beta1.TCPIngress:
+		return updateTCPIngressReferredSecrets(ctx, client, dataplaneClient, obj)
+	}
+
+	return nil
+}
```

Comment:
Do you think it would make sense to introduce more separation of concerns here and split extracting referred objects and updating them? This would mean having a function per kind that extracts referred objects, similarly to `listSecretsReferredByGateway`, making it easily unit testable. IMO it would be worth having such extraction logic covered with tests.

Reply:
Probably that would also reduce redundancy - updating logic is more or less the same for every kind as I can see (with `Ingress` exception that requires objects to exist in k8s).

Reply:
Thanks for your advice. Updated to the format you mentioned here.